xml/api/ticket-2.34.rng

```diff
@@ -0,0 +1,60 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<grammar xmlns="http://relaxng.org/ns/structure/1.0"
+         datatypeLibrary="http://www.w3.org/2001/XMLSchema-datatypes">
+
+    <include href="../constraints-3.9.rng">
+        <start>
+            <ref name="element-ticket"/>
+        </start>
+
+        <!-- This redefines the constraints element from constraints-X.X.rng
+             so it can only contain the element-rsc_ticket element.  This allows
+             us to restrict it so only those kinds of constraints are allowed
+             in ticket XML.
+          -->
+        <define name="element-constraints">
+            <element name="constraints">
+                <zeroOrMore>
+                    <ref name="element-rsc_ticket"/>
+                </zeroOrMore>
+            </element>
+        </define>
+    </include>
+
+    <define name="element-ticket">
+        <element name="ticket">
+            <attribute name="id"> <data type="ID" /> </attribute>
+            <optional>
+                <attribute name="status">
+                    <choice>
+                        <value>granted</value>
+                        <value>revoked</value>
+                    </choice>
+                </attribute>
+            </optional>
+            <optional>
+                <attribute name="standby"> <data type="boolean" /> </attribute>
+            </optional>
+            <optional>
+                <attribute name="last-granted"> <text /> </attribute>
+            </optional>
+            <optional>
+                <ref name="element-constraints" />
+            </optional>
+            <optional>
+                <element name="ticket_state">
+                    <zeroOrMore>
+                        <attribute> <anyName /> <text /> </attribute>
+                    </zeroOrMore>
+                </element>
+            </optional>
+            <optional>
+                <element name="attribute">
+                    <attribute name="name"> <text /> </attribute>
+                    <attribute name="value"> <text /> </attribute>
+                </element>
+            </optional>
+        </element>
+    </define>
```

Comment:
@kgaillot Looking at the XML files in cts/scheduler/xml, it looks like there's only ever `<ticket_state>` elements, and that those elements contain id, granted, last-granted, standby, etc.  And then looking at `modify_ticket_state` in crm_ticket.c, it looks like any of the attributes that could be added to a ticket would also go into the `<ticket_state>`.

So with that in mind, I feel like perhaps the schema here is redundant.  The fact that crm_mon is already using a ticket schema makes it a little more complicated, but maybe the right thing to do is simply remove the `<ticket_state>` portion.  I think everything in there is covered by the higher up element.  I think I'd need to move the generic attribute stuff up a level, too.

One thing that's confusing me a little bit here is that in the CIB, it's always `<ticket_state>`, but in the API, it would be `<ticket>`.  Maybe that's only confusing on the implementation side, and it wouldn't be a problem for users.

Reply:
Sounds good. The CIB has both `rsc_ticket` in constraints and `ticket_state` in state. I don't think it's a problem that the API uses something different to present the info to the user.